ios/RCTOneSignal/RCTOneSignalEventEmitter.m

```diff
@@ -173,6 +154,42 @@ + (void)sendEventWithName:(NSString *)name withBody:(NSDictionary *)body {
     }
 }
 
+RCT_REMAP_METHOD(getDeviceState,
+                getDeviceStateResolver:(RCTPromiseResolveBlock)resolve
+                rejecter:(RCTPromiseRejectBlock)reject) {
+    OSDeviceState *deviceState = [OneSignal getDeviceState];
+
+    if (deviceState) {
+        resolve([deviceState jsonRepresentation]);
+    } else {
+        NSError * error = nil;
+        NSString * message = @"Could not get OneSignal device state";
+        reject(@"no_value", message, error);
+    }
+}
+
+RCT_EXPORT_METHOD(setNotificationOpenedHandler) {
+    [OneSignal setNotificationOpenedHandler:^(OSNotificationOpenedResult *result) {
+        [RCTOneSignalEventEmitter sendEventWithName:@"OneSignal-remoteNotificationOpened" withBody:[result jsonRepresentation]];
+    }];
+}
+
+RCT_EXPORT_METHOD(setNotificationWillShowInForegroundHandler) {
+    [OneSignal setNotificationWillShowInForegroundHandler:^(OSNotification *notif, OSNotificationDisplayResponse completion) {
+        self->receivedNotificationCache[notif.notificationId] = notif;
+        self->notificationCompletionCache[notif.notificationId] = completion;
+        [RCTOneSignalEventEmitter sendEventWithName:@"OneSignal-notificationWillShowInForeground" withBody:[notif jsonRepresentation]];
+    }];
+}
+
+RCT_EXPORT_METHOD(completeNotificationJob:(NSString*)notificationId) {
```

Comment:
This method needs to take a boolean for silencing the notification or not. The way the complete method works is that passing a notification the completion handler means it will show, and passing nil to the handler means it will be silenced. We can't just pass nil for the notificationId to this method because then we can't get the completion block. If we pass in a boolean from the javascript we can use that to decide if we provide a notification to the completion block or just pass nil.